core/offchain/impl/offchain_worker_impl.cpp

```diff
@@ -0,0 +1,340 @@
+        primitives::BlockInfo(header_.number, hash);
+
+    local_storage_ =
+        std::make_shared<OffchainLocalStorageImpl>(std::move(storage));
+  }
+
+  outcome::result<void> OffchainWorkerImpl::run() {
+    BOOST_ASSERT(not current().has_value());
+
+    auto main_thread_func = [ocw = shared_from_this()] {
+      soralog::util::setThreadName("ocw." + std::to_string(ocw->block_.number));
+
+      current(*ocw);
+
+      SL_TRACE(ocw->log_,
+               "Offchain worker is started for block #{} hash={}",
+               ocw->block_.number,
+               ocw->block_.hash.toHex());
+
+      auto res = ocw->executor_->callAt<void>(
+          ocw->block_.hash, "OffchainWorkerApi_offchain_worker", ocw->header_);
+
+      current(std::nullopt);
+
+      if (res.has_failure()) {
+        SL_ERROR(ocw->log_,
+                 "Can't execute offchain worker for block #{} hash={}: {}",
+                 ocw->block_.number,
+                 ocw->block_.hash.toHex(),
+                 res.error().message());
+        return;
+      }
+
+      SL_DEBUG(ocw->log_,
+               "Offchain worker is successfully executed for block #{} hash={}",
+               ocw->block_.number,
+               ocw->block_.hash.toHex());
+    };
+
+    try {
+      std::thread(std::move(main_thread_func)).detach();
+    } catch (const std::system_error &exception) {
+      return outcome::failure(exception.code());
+    } catch (...) {
+      return outcome::failure(boost::system::error_code{});
+    }
+
+    return outcome::success();
+  }
+
+  bool OffchainWorkerImpl::isValidator() const {
+    bool isValidator = app_config_.roles().flags.authority == 1;
+    return isValidator;
+  }
+
+  Result<Success, Failure> OffchainWorkerImpl::submitTransaction(
+      const primitives::Extrinsic &ext) {
+    auto result = author_api_->submitExtrinsic(ext);
+    if (result.has_value()) {
+      return Success();
+    }
+    return Failure();
+  }
+
+  Result<OpaqueNetworkState, Failure> OffchainWorkerImpl::networkState() {
+    auto peer_info = host_.getPeerInfo();
+    OpaqueNetworkState result{.peer_id = peer_info.id};
+
+    std::list<libp2p::multi::Multiaddress> address(peer_info.addresses.begin(),
+                                                   peer_info.addresses.end());
+
+    return result;
+  }
+
+  Timestamp OffchainWorkerImpl::timestamp() {
+    return std::chrono::duration_cast<std::chrono::milliseconds>(
+               clock_->now().time_since_epoch())
+        .count();
+  }
+
+  void OffchainWorkerImpl::sleepUntil(Timestamp deadline) {
+    auto ts = clock_->zero() + std::chrono::milliseconds(deadline);
+    SL_TRACE(log_,
+             "Falling asleep till {} (for {}ms)",
+             deadline,
+             std::chrono::duration_cast<std::chrono::milliseconds>(
+                 ts - clock_->now())
+                 .count());
+
+    std::this_thread::sleep_until(std::chrono::system_clock::time_point()
```

Comment:
time_point can be constructed from duration.

Reply:
Could you show example?